File: src/omnia/data/structure/mutable/LinkedQueue.java
package omnia.data.structure.mutable;

import static java.util.Objects.requireNonNull;

import java.util.Optional;

public final class LinkedQueue<E> implements Queue<E> {

  private final MutableList<E> items = new LinkedList<>();

  @Override
  public void enqueue(E item) {
    items.add(requireNonNull(item));
  }

  @Override
  public Optional<E> dequeue() {
    if (!items.isPopulated()) {
      return Optional.empty();
    }
    E item = items.itemAt(0);
    items.removeAt(0);
    return Optional.of(item);
  }

  @Override
  public Optional<E> peek() {
    return items.isPopulated() ? Optional.of(items.itemAt(0)) : Optional.empty();
  }

  @Override
  public int count() {
    return items.count();
  }

  @Override
  public boolean isPopulated() {
    return items.isPopulated();
  }
}

File: src/omnia/data/structure/mutable/LinkedQueue.kt
package omnia.data.structure.mutable

import java.util.Objects
import java.util.Optional

class LinkedQueue<E: Any> : Queue<E> {
    private val items: MutableList<E> = LinkedList()
    override fun enqueue(item: E) {
        items.add(Objects.requireNonNull(item))
    }

    override fun dequeue(): Optional<E> {
        if (!items.isPopulated) {
            return Optional.empty()
        }
        val item = items.itemAt(0)
        items.removeAt(0)
        return Optional.of(item)
    }

    override fun peek(): Optional<E> {
        return if (items.isPopulated) Optional.of(items.itemAt(0)) else Optional.empty()
    }

    override fun count(): Int {
        return items.count()
    }

    override val isPopulated: Boolean
        get() = items.isPopulated
}